src/nix/nix_store.rs

```diff
@@ -84,20 +90,24 @@ impl NixStoreCmd {
     }
 
     /// Return the derivation used to build the given build output.
-    async fn nix_store_query_deriver(&self, out_path: PathBuf) -> Result<DrvOut> {
+    async fn nix_store_query_deriver(&self, out_path: PathBuf) -> Result<DrvOut, NixCmdError> {
         let mut cmd = self.command();
         cmd.args(["--query", "--deriver", &out_path.to_string_lossy().as_ref()]);
         nix_rs::command::trace_cmd(&cmd);
-        let out = cmd.output().await?;
+        let out = cmd
+            .output()
+            .await
+            .map_err(CommandError::ChildProcessError)?;
         if out.status.success() {
             let drv_path = String::from_utf8(out.stdout)?.trim().to_string();
+            if drv_path == "unknown-deriver" {
+                return Err(NixCmdError::UnknownDeriverError);
+            }
             Ok(DrvOut(PathBuf::from(drv_path)))
         } else {
-            let exit_code = out.status.code().unwrap_or(1);
-            bail!(
-                "nix-store --query --deriver failed to run (exited: {})",
-                exit_code
-            );
+            let stderr = String::from_utf8(out.stderr).ok();
+            let exit_code = out.status.code();
+            Err(CommandError::ProcessFailed { stderr, exit_code }.into())
```

Comment:
There's a bug there. If UTF8 decoding fails, this sets stderr to `None` thus creating a false impression that no stderr is missing. Just use lossy decoding.